### ISSUE TITLE
Common UI improvements

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#5392]
Goals for UI:

1. Abstract UI implementation to apply high-level functions/classes to all UI components
  - [x] #5347 and #5348 
2. Establish CLI/UI guidelines

3. Improve formatting, style, and functionality

@skshetry @efiop @pmrowla

**Comment by dberenbaum** [comment]
See #5357 and #5348 for abstraction

**Comment by dberenbaum** [comment]
Referenced in https://github.com/iterative/dvc/pull/5282#discussion_r570719275: `Subcommand should usually be plural`

**Comment by jorgeorpinel** [comment]
What about extracting all user messages to a central location for easy access to the prod team to help with? Could even help for future i18n ([more context](https://github.com/iterative/dvc.org/issues/2108#issuecomment-774580721)).

**Comment by dberenbaum** [comment]
See https://github.com/iterative/dvc/discussions/5891 for `--quiet` guidelines

**Comment by dberenbaum** [comment]
@skshetry Are there other issues that can be linked here?

**Comment by skshetry** [comment]
Some update on this - We decided to adopt https://clig.dev as a base guideline for the CLI applications (I would suggest everyone read this). I started going through our commands and fix where we violate this and make some improvements along the way. 

I had to start somewhere, so I have chosen data management commands - starting with `add`.  I noticed some performance and reliability issues on `add`: #6227, #6226, and #6228, which I'll be going to fix, but before I need to do some cleanups.

After that, I'll focus on improving data management commands (`add/commit/remove/move/checkout/destroy`).

**Comment by skshetry** [comment]
Another thing that I'd like to open the discussion is the UI messages for the `add`.

- The `add` command is quiet about what is done, at the end, there's only a progress bar (and, an ugly hint). What should it say on success?
- Do we even need a progress bar for a single target to add (eg: `dvc add data`)?
- Do we need progress bars at all, even for multiple targets?
- How should DVC show progress bars/status on places where it might take a lot of time? eg: recursively adding a large directory, collecting lots of stages, etc? 

To compare (and, spark the discussion, here are the asciinema recordings for multiple targets add and single targets add).

#### Single target
[![asciicast](https://asciinema.org/a/EMl7w6cAF27eobziIXKUC23UV.svg)](https://asciinema.org/a/EMl7w6cAF27eobziIXKUC23UV)

#### Multiple targets
[![asciicast](https://asciinema.org/a/xO4lQyKN1Uyl3h8D909pLpMRu.svg)](https://asciinema.org/a/xO4lQyKN1Uyl3h8D909pLpMRu)

**Comment by dberenbaum** [comment]
Another guide for reference: https://primer.style/cli/

**Comment by jorgeorpinel** [comment]
> We decided to adopt https://clig.dev as a base guideline for the CLI applications

p.s. would it be worth giving a quick talk to others on this? E.g. in an all hands meeting or in a bash perhaps (cc @casperdcl)

**Comment by jorgeorpinel** [comment]
> The add command is quiet about what is done, at the end, there's only a progress bar (and, an ugly hint).

Not sure why you consider the progress bar to be quiet. I think it communicates what's happening clearly enough.

💡 The exact phrases used in all messages can always be improved and to that effect I propose to consider an i18n-like effort to refactor the code extracting message strings to a separate module or config file.

Idk what to think about the Git hints. I give them for granted at this point but thinking about it, I'm not sure. Are they useful realistically? Or should people know what to do after `dvc add`? Maye changing that to a more explicit message on what happened would be better e.g. `'data.csv' was cached. './data.csv.dvc' was created.`

> Do we need progress bars at all

Yes. These operations can take a long time. It's important to know that there's progress.

> How should DVC show progress bars/status on places where it might take a lot of time?

I see some problems with the current way (shown in the 2nd animation above):
1. The main progress bar starts at `0/2`. Why 0? And it never prints the name `data` (first dir given)
2. "computing hashes" states is `only done once` but it seems to happen more than once (once per directory I guess).
3. The 2nd top bar jumps from 0 to `2/2` and prints the name `data2` (2nd dir given).

In all it is a bit confusing to try interpret those bars. Can there be a single, persistent master bar on top and then smaller bars for each subprocess underneath?

**Comment by iesahin** [comment]
> I give them for granted at this point but thinking about it, I'm not sure. Are they useful realistically? Or should people know what to do after `dvc add`?

They are misleading. I think silence is better than providing an incorrect command. (Some of these files are `.gitignore`d, see the difference between proposed `git add` and `git status`)

![image](https://user-images.githubusercontent.com/476310/129159598-21debe47-3d47-4ce2-871e-37d5b2801fe7.png)

I believe building the command verbose doesn't add much value anyway. For the first few times, they may lead the user but after that, they are ignored. Verbosity must be optional, or at least could be turned off using config.

**Comment by skshetry** [comment]
> > Or should people know what to do after `dvc add`?

I am hesitant to remove this on `dvc add` as I have seen a lot of users just copy and paste that. I don't expect everyone using DVC to understand git. They might not get very far, but for "Data Versioning" in DVC, it's enough. And, `git add` hint on `dvc add` is very stable. I do have plans for suppressing those hints messages.

For experiments and pipelines, I don't have strong opinions.

**Comment by skshetry** [comment]
Update on UI improvement:
1. Index was introduced in #6300.
2. In `dvc add`, some progress bars were merged and reduced. A few spinners during stage collection and graph checks were added. We still need to fix how we save files to cache which we decided on https://github.com/iterative/dvc/discussions/6252.
   Also see https://github.com/iterative/dvc/issues/6387.
3. `dvc add` duplicated targets are now ignored, instead of failing with ambiguous error.
4. Updater was changed to only show on `dvc version`, thanks to @efiop.
5. Updater message was changed from in-box to `pip` like output.
6. Introduced DVC_ANALYTICS env var thanks to @efiop.
7. fsspec callback was introduced by @isidentical and `push`/`pull` transfers were unified by @pmrowla, making it possible for us to have a single progress bar: https://github.com/iterative/dvc/issues/3682.

The #3682 is now being worked on. 🙂
